File: tools/detect_tag_changes.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "json"
require "optparse"
require "set"

# Script to detect changes in normative tags extracted from asciidoc files
# Compares two tag JSON files and reports additions, deletions, and modifications

class TagChanges
  attr_reader :added, :deleted, :modified

  def initialize
    @added = {}      # Tags present in current but not in reference
    @deleted = {}    # Tags present in reference but not in current
    @modified = {}   # Tags present in both but with different text
  end

  def any_changes?
    !@added.empty? || !@deleted.empty? || !@modified.empty?
  end

  def total_changes
    @added.size + @deleted.size + @modified.size
  end
end

class TagChangeDetector
  def initialize(options = {})
    @verbose = options[:verbose] || false
  end

  # Load tags from a JSON file
  # @param filename [String] Path to the JSON file
  # @return [Hash<String, String>] Hash of tag names to tag text
  def load_tags(filename)
    unless File.exist?(filename)
      abort("Error: File not found: #{filename}")
    end

    begin
      data = JSON.parse(File.read(filename))
      tags = data["tags"] || {}
      tags
    rescue JSON::ParserError => e
      abort("Error: Failed to parse JSON from #{filename}: #{e.message}")
    end
  end

  # Compare two tag sets and identify changes
  # @param reference_tags [Hash<String, String>] Original tags
  # @param current_tags [Hash<String, String>] Updated tags
  # @return [TagChanges] Object containing all changes
  def detect_changes(reference_tags, current_tags)
    changes = TagChanges.new

    reference_keys = reference_tags.keys.to_set
    current_keys = current_tags.keys.to_set

    # Find added tags (in current but not in reference)
    added_keys = current_keys - reference_keys
    added_keys.each do |tag_name|
      changes.added[tag_name] = current_tags[tag_name]
    end

    # Find deleted tags (in reference but not in current)
    deleted_keys = reference_keys - current_keys
    deleted_keys.each do |tag_name|
      changes.deleted[tag_name] = reference_tags[tag_name]
    end

    # Find modified tags (in both but different text)
    common_keys = reference_keys & current_keys
    common_keys.each do |tag_name|
      reference_text = reference_tags[tag_name]
      current_text = current_tags[tag_name]

      # Compare normalized text (ignoring whitespace and AsciiDoc formatting differences)
      normalized_ref = normalize_text(reference_text)
      normalized_cur = normalize_text(current_text)

      if normalized_ref != normalized_cur
        changes.modified[tag_name] = {
          "reference" => reference_text,
          "current" => current_text
        }
      end
    end

    changes
  end

  # Format and display changes
  # @param changes [TagChanges] Changes to display
  # @param reference_file [String] Name of reference file (for display)
  # @param current_file [String] Name of current file (for display)
  # @param verbose [Boolean] Whether to show verbose output
  def display_changes(changes, reference_file, current_file, verbose)
    if verbose
      puts "=" * 80
      puts "Tag Changes Report"
      puts "=" * 80
      puts
    end

    puts "Reference file: #{reference_file}"
    puts "Current file: #{current_file}"

    unless changes.any_changes?
      puts "No changes detected."
      return
    end

    # Display added tags
    unless changes.added.empty?
      count = changes.added.size
      puts "Added #{count} tag#{count > 1 ? 's' : ''}:"
      changes.added.sort.each do |tag_name, text|
        puts %Q[  * "#{tag_name}": "#{truncate_text(text)}"]
      end
      puts
    end

    # Display deleted tags
    unless changes.deleted.empty?
      count = changes.deleted.size
      puts "Deleted #{count} tag#{count > 1 ? 's' : ''}:"
      changes.deleted.sort.each do |tag_name, text|
        puts %Q[  * "#{tag_name}": "#{truncate_text(text)}"]
      end
      puts
    end

    # Display modified tags
    unless changes.modified.empty?
      count = changes.modified.size
      puts "Modified #{count} tag#{count > 1 ? 's' : ''}:"
      changes.modified.sort.each do |tag_name, texts|
        puts %Q[  * "#{tag_name}":]
        puts %Q[      Reference: "#{truncate_text(texts['reference'])}"]
        puts %Q[      Current:   "#{truncate_text(texts['current'])}"]
      end
      puts
    end

    # Summary
    if verbose
      puts "=" * 80
      puts "Summary: #{changes.total_changes} total changes"
      puts "  Added:    #{changes.added.size}"
      puts "  Deleted:  #{changes.deleted.size}"
      puts "  Modified: #{changes.modified.size}"
      puts "=" * 80
    end
  end


  # Update a tags file by adding new tags from additions
  # @param file_path [String] Path to the file to update
  # @param changes [TagChanges] Changes detected
  def update_tags_file(file_path, changes)
    if changes.added.empty?
      puts "No additions to merge into #{file_path}"
      puts "Skipping file update - no additions" if @verbose
      return
    end

    unless File.exist?(file_path)
      abort("Error: Cannot update file - not found: #{file_path}")
    end

    puts "Updating reference file: #{file_path}" if @verbose

    begin
      data = JSON.parse(File.read(file_path))
      original_count = data["tags"].size

      # Add new tags
      changes.added.each do |tag_name, tag_text|
        puts "  Adding tag: #{tag_name}" if @verbose
        data["tags"][tag_name] = tag_text
      end

      # Write back to file
      puts "Writing updated file..." if @verbose
      File.write(file_path, JSON.pretty_generate(data))
      new_count = data["tags"].size
      puts "Updated #{file_path}: added #{changes.added.size} new tags (#{original_count} -> #{new_count} total tags)"
    rescue JSON::ParserError => e
      abort("Error: Failed to parse JSON from #{file_path}: #{e.message}")
    end
  end

  private

  # Normalize text for comparison (whitespace and AsciiDoc formatting)
  # @param text [String] Text to normalize
  # @return [String] Normalized text
  def normalize_text(text)
    # First normalize whitespace, then strip AsciiDoc formatting
    strip_asciidoc_formatting(normalize_whitespace(text))
  end

  # Normalize whitespace for comparison
  # @param text [String] Text to normalize
  # @return [String] Text with normalized whitespace
  def normalize_whitespace(text)
    text.strip.gsub(/\s+/, ' ')
  end

  # Strip AsciiDoc formatting marks
  # @param text [String] Text to strip formatting from
  # @return [String] Text without AsciiDoc formatting
  def strip_asciidoc_formatting(text)
    result = text.dup

    # Remove bold: **text** (unconstrained) or *text* (constrained)
    result.gsub!(/\*\*([^\*]+?)\*\*/, '\1')
    result.gsub!(/\*([^\*]+?)\*/, '\1')

    # Remove italic: __text__ (unconstrained) or _text_ (constrained, not in middle of words)
    result.gsub!(/__([^_]+?)__/, '\1')
    result.gsub!(/(?<!\w)_([^_]+?)_(?!\w)/, '\1')

    # Remove monospace: `text`
    result.gsub!(/`([^`]+?)`/, '\1')

    # Remove superscript: ^text^
    result.gsub!(/\^([^\^]+?)\^/, '\1')

    # Remove subscript: ~text~
    result.gsub!(/~([^~]+?)~/, '\1')

    # Remove role-based formatting: [role]#text#
    result.gsub!(/\[[^\]]+\]#([^#]+?)#/, '\1')

    # Remove cross-references: <<anchor,text>> or <<anchor>>
    result.gsub!(/&lt;&lt;[^,&]+,([^&]+)&gt;&gt;/, '\1')
    result.gsub!(/&lt;&lt;[^&]+&gt;&gt;/, '')

    # Remove passthrough: +++text+++
    result.gsub!(/\+\+\+([^\+]+?)\+\+\+/, '\1')

    result.strip
  end

  # Truncate text for display
  # @param text [String] Text to truncate
  # @param max_length [Integer] Maximum length before truncation
  # @return [String] Truncated text
  def truncate_text(text, max_length = 100)
    return text if text.length <= max_length
    "#{text[0...max_length]}..."
  end
end

# Parse command-line arguments
def parse_options
  options = {
    update_reference: false,
    verbose: false
  }

  parser = OptionParser.new do |opts|
    opts.banner = "Usage: #{File.basename($0)} [options] REFERENCE_TAGS.json CURRENT_TAGS.json"
    opts.separator ""
    opts.separator "Detect changes in normative tags between two JSON files"
    opts.separator ""
    opts.separator "Options:"

    opts.on("-u", "--update-reference", "Update the reference tags file by adding any additions found in the current file") do
      options[:update_reference] = true
    end

    opts.on("-v", "--verbose", "Show verbose output with detailed processing information") do
      options[:verbose] = true
    end

    opts.on("-h", "--help", "Show this help message") do
      puts opts
      exit
    end
  end

  parser.parse!

  if ARGV.length != 2
    puts parser
    exit 1
  end

  options[:reference_file] = ARGV[0]
  options[:current_file] = ARGV[1]

  options
end

# Main execution
if __FILE__ == $0
  options = parse_options

  detector = TagChangeDetector.new(
    verbose: options[:verbose]
  )

  # Load both tag files
  reference_tags = detector.load_tags(options[:reference_file])
  current_tags = detector.load_tags(options[:current_file])

  # Detect changes
  changes = detector.detect_changes(reference_tags, current_tags)

  # Display changes
  detector.display_changes(changes, options[:reference_file], options[:current_file], options[:verbose])

  # Update reference file if requested
  if options[:update_reference]
    detector.update_tags_file(options[:reference_file], changes)
  end

  # Exit with appropriate status code
  # Return 0 if no changes or only additions
  # Return 1 if any modifications or deletions detected
  has_modifications_or_deletions = !changes.modified.empty? || !changes.deleted.empty?
  exit(has_modifications_or_deletions ? 1 : 0)
end

File: tools/detect_tag_changes.py
#!/usr/bin/env python3
"""Script to detect changes in normative tags extracted from asciidoc files.

Compares two tag JSON files and reports additions, deletions, and modifications.
"""

import json
import sys
import argparse
import re
from pathlib import Path
from typing import Dict, Set, Any


class TagChanges:
    """Container for tag change information."""

    def __init__(self):
        self.added: Dict[str, str] = {}      # Tags present in current but not in reference
        self.deleted: Dict[str, str] = {}    # Tags present in reference but not in current
        self.modified: Dict[str, Dict[str, str]] = {}   # Tags present in both but with different text

    def any_changes(self) -> bool:
        """Check if any changes were detected."""
        return bool(self.added or self.deleted or self.modified)

    def total_changes(self) -> int:
        """Get total number of changes."""
        return len(self.added) + len(self.deleted) + len(self.modified)


class TagChangeDetector:
    """Detector for changes in tag files."""

    def __init__(self, verbose: bool = False, strict: bool = False):
        self.verbose = verbose
        self.strict = strict

    def load_tags(self, filename: str) -> Dict[str, str]:
        """Load tags from a JSON file.

        Args:
            filename: Path to the JSON file

        Returns:
            Dict of tag names to tag text
        """
        file_path = Path(filename)
        if not file_path.exists():
            sys.exit(f"Error: File not found: {filename}")

        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            tags = data.get("tags", {})
            return tags
        except json.JSONDecodeError as e:
            sys.exit(f"Error: Failed to parse JSON from {filename}: {e}")
        except Exception as e:
            sys.exit(f"Error: Failed to read {filename}: {e}")

    def detect_changes(self, reference_tags: Dict[str, str],
                      current_tags: Dict[str, str]) -> TagChanges:
        """Compare two tag sets and identify changes.

        Args:
            reference_tags: Original tags
            current_tags: Updated tags

        Returns:
            TagChanges object containing all changes
        """
        changes = TagChanges()

        reference_keys: Set[str] = set(reference_tags.keys())
        current_keys: Set[str] = set(current_tags.keys())

        # Find added tags (in current but not in reference)
        added_keys = current_keys - reference_keys
        for tag_name in added_keys:
            changes.added[tag_name] = current_tags[tag_name]

        # Find deleted tags (in reference but not in current)
        deleted_keys = reference_keys - current_keys
        for tag_name in deleted_keys:
            changes.deleted[tag_name] = reference_tags[tag_name]

        # Find modified tags (in both but different text)
        common_keys = reference_keys & current_keys
        for tag_name in common_keys:
            reference_text = reference_tags[tag_name]
            current_text = current_tags[tag_name]

            # Compare normalized text (ignoring whitespace and AsciiDoc formatting differences)
            normalized_ref = self._normalize_text(reference_text)
            normalized_cur = self._normalize_text(current_text)

            if normalized_ref != normalized_cur:
                changes.modified[tag_name] = {
                    "reference": reference_text,
                    "current": current_text
                }

        return changes

    def display_changes(self, changes: TagChanges, reference_file: str,
                       current_file: str, verbose: bool):
        """Format and display changes.

        Args:
            changes: Changes to display
            reference_file: Name of reference file (for display)
            current_file: Name of current file (for display)
            verbose: Whether to show verbose output
        """
        if verbose:
            print("=" * 80)
            print("Tag Changes Report")
            print("=" * 80)
            print()

        print(f"Reference file: {reference_file}")
        print(f"Current file: {current_file}")

        if not changes.any_changes():
            print("No changes detected.")
            return

        # Display added tags
        if changes.added:
            count = len(changes.added)
            print(f"Added {count} tag{'s' if count > 1 else ''}:")
            for tag_name in sorted(changes.added.keys()):
                text = changes.added[tag_name]
                print(f'  * "{tag_name}": "{self._truncate_text(text)}"')
            print()

        # Display deleted tags
        if changes.deleted:
            count = len(changes.deleted)
            print(f"Deleted {count} tag{'s' if count > 1 else ''}:")
            for tag_name in sorted(changes.deleted.keys()):
                text = changes.deleted[tag_name]
                print(f'  * "{tag_name}": "{self._truncate_text(text)}"')
            print()

        # Display modified tags
        if changes.modified:
            count = len(changes.modified)
            print(f"Modified {count} tag{'s' if count > 1 else ''}:")
            for tag_name in sorted(changes.modified.keys()):
                texts = changes.modified[tag_name]
                print(f'  * "{tag_name}":')
                print(f'      Reference: "{self._truncate_text(texts["reference"])}"')
                print(f'      Current:   "{self._truncate_text(texts["current"])}"')
            print()

        # Summary
        if verbose:
            print("=" * 80)
            print(f"Summary: {changes.total_changes()} total changes")
            print(f"  Added:    {len(changes.added)}")
            print(f"  Deleted:  {len(changes.deleted)}")
            print(f"  Modified: {len(changes.modified)}")
            print("=" * 80)

    def update_tags_file(self, file_path: str, changes: TagChanges):
        """Update a tags file by adding new tags from additions.

        Args:
            file_path: Path to the file to update
            changes: Changes detected
        """
        if not changes.added:
            print(f"No additions to merge into {file_path}")
            if self.verbose:
                print("Skipping file update - no additions")
            return

        path = Path(file_path)
        if not path.exists():
            sys.exit(f"Error: Cannot update file - not found: {file_path}")

        if self.verbose:
            print(f"Updating reference file: {file_path}")

        try:
            with open(path, 'r', encoding='utf-8') as f:
                data = json.load(f)

            original_count = len(data["tags"])

            # Add new tags
            for tag_name, tag_text in changes.added.items():
                if self.verbose:
                    print(f"  Adding tag: {tag_name}")
                data["tags"][tag_name] = tag_text

            # Write back to file
            if self.verbose:
                print("Writing updated file...")

            with open(path, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2, ensure_ascii=False)

            new_count = len(data["tags"])
            print(f"Updated {file_path}: added {len(changes.added)} new tags ({original_count} -> {new_count} total tags)")

        except json.JSONDecodeError as e:
            sys.exit(f"Error: Failed to parse JSON from {file_path}: {e}")
        except Exception as e:
            sys.exit(f"Error: Failed to update {file_path}: {e}")

    def _normalize_text(self, text: str) -> str:
        """Normalize text for comparison.

        In strict mode, only whitespace is normalized (line-ending portability).
        Otherwise, AsciiDoc formatting marks are also stripped, so e.g. wrapping
        a term in `monospace` does not count as a modification.

        Args:
            text: Text to normalize

        Returns:
            Normalized text
        """
        normalized = self._normalize_whitespace(text)
        if self.strict:
            return normalized
        return self._strip_asciidoc_formatting(normalized)

    def _normalize_whitespace(self, text: str) -> str:
        """Normalize whitespace for comparison.

        Args:
            text: Text to normalize

        Returns:
            Text with normalized whitespace
        """
        return re.sub(r'\s+', ' ', text.strip())

    def _normalize_riscv_macros(self, text: str) -> str:
        """Normalize RISC-V AsciiDoc macros to their plain-text equivalent.

        The tags backend stores raw AsciiDoc (see #169), so a pure markup
        migration such as ``mip`` -> ``csr:mip[]`` must not be reported as a
        change to normative text. These are the semantic macros used across
        the ISA manual:

            csr:NAME[]        -> NAME           (csr:mip[]         -> mip)
            csr:NAME[FIELD]   -> NAME.FIELD     (csr:vsstatus[sdt] -> vsstatus.SDT)
            csr:[FIELD]       -> FIELD          (csr:[tm]          -> TM)
            csr::[FIELD]      -> FIELD          (csr::[fs]         -> FS)
            ext:NAME[]        -> NAME           (ext:f[]           -> F)
            insn:NAME[]       -> NAME           (insn:wrs.sto[]    -> wrs.sto)

        CSR names keep their case; CSR field names and extension names render
        upper-case, matching how they read as prose.

        Note this normalization is for the reviewer-facing change *report*
        only. It deliberately erases formatting differences, which is correct
        for "did the normative text change" and wrong for "is this file byte
        up to date" -- do not reuse it as a freshness check.

        Args:
            text: Text to normalize

        Returns:
            Text with RISC-V macros reduced to plain text
        """
        def _csr_named(match: "re.Match") -> str:
            name, field = match.group(1), match.group(2)
            return f"{name}.{field.upper()}" if field else name

        # csr:NAME[FIELD] and csr:NAME[] (name present, field optional).
        result = re.sub(r'csr:([\w.]+)\[([\w.]*)\]', _csr_named, text)
        # csr:[FIELD] and csr::[FIELD] (name absent).
        result = re.sub(r'csr::?\[([\w.]+)\]',
                        lambda m: m.group(1).upper(), result)
        # ext:NAME[] -> NAME (capitalized as prose does: single-letter base
        # extensions read as "F"/"D"; named ones as "Sscofpmf"/"Zicsr").
        result = re.sub(r'ext:([\w.]+)\[\]',
                        lambda m: m.group(1).capitalize(), result)
        # insn:NAME[] -> NAME.
        result = re.sub(r'insn:([\w.]+)\[\]', r'\1', result)

        # Binary renotation: normalize both 11b and 0b11 spellings to 0b-prefixed.
        result = re.sub(r'\b([01]+)b\b', r'0b\1', result)

        # En dash: -- -> {endash} (leave em dash --- untouched).
        result = re.sub(r'(?<!-)--(?!-)', '{endash}', result)

        return result

    def _strip_asciidoc_formatting(self, text: str) -> str:
        """Strip AsciiDoc formatting marks.

        Args:
            text: Text to strip formatting from

        Returns:
            Text without AsciiDoc formatting
        """
        # Reduce RISC-V semantic macros first so a macro wrapped in other
        # formatting is still recognized before the generic strips run.
        result = self._normalize_riscv_macros(text)

        # Remove bold: **text** (unconstrained) or *text* (constrained)
        result = re.sub(r'\*\*([^\*]+?)\*\*', r'\1', result)
        result = re.sub(r'\*([^\*]+?)\*', r'\1', result)

        # Remove italic: __text__ (unconstrained) or _text_ (constrained, not in middle of words)
        result = re.sub(r'__([^_]+?)__', r'\1', result)
        result = re.sub(r'(?<!\w)_([^_]+?)_(?!\w)', r'\1', result)

        # Remove monospace: `text`
        result = re.sub(r'`([^`]+?)`', r'\1', result)

        # Remove superscript: ^text^
        result = re.sub(r'\^([^\^]+?)\^', r'\1', result)

        # Remove subscript: ~text~
        result = re.sub(r'~([^~]+?)~', r'\1', result)

        # Remove role-based formatting: [role]#text#
        result = re.sub(r'\[[^\]]+\]#([^#]+?)#', r'\1', result)

        # Remove cross-references: <<anchor,text>> or <<anchor>>
        result = re.sub(r'&lt;&lt;[^,&]+,([^&]+)&gt;&gt;', r'\1', result)
        result = re.sub(r'&lt;&lt;[^&]+&gt;&gt;', '', result)

        # Remove passthrough: +++text+++
        result = re.sub(r'\+\+\+([^\+]+?)\+\+\+', r'\1', result)

        return result.strip()

    def _truncate_text(self, text: str, max_length: int = 100) -> str:
        """Truncate text for display.

        Args:
            text: Text to truncate
            max_length: Maximum length before truncation

        Returns:
            Truncated text
        """
        if len(text) <= max_length:
            return text
        return f"{text[:max_length]}..."


def parse_options() -> argparse.Namespace:
    """Parse command-line arguments.

    Returns:
        Namespace containing parsed arguments
    """
    parser = argparse.ArgumentParser(
        description="Detect changes in normative tags between two JSON files",
        formatter_class=argparse.RawDescriptionHelpFormatter
    )

    parser.add_argument(
        'reference_file',
        help='Path to the reference tags JSON file'
    )

    parser.add_argument(
        'current_file',
        help='Path to the current tags JSON file'
    )

    parser.add_argument(
        '-u', '--update-reference',
        action='store_true',
        help='Update the reference tags file by adding any additions found in the current file'
    )

    parser.add_argument(
        '-s', '--strict',
        action='store_true',
        help=('Treat additions as failures and compare prose byte-for-byte '
              '(only whitespace is normalized). Use this in CI when the '
              'reference file must match the build output exactly.')
    )

    parser.add_argument(
        '-v', '--verbose',
        action='store_true',
        help='Show verbose output with detailed processing information'
    )

    return parser.parse_args()


def main():
    """Main execution."""
    options = parse_options()

    detector = TagChangeDetector(verbose=options.verbose, strict=options.strict)

    # Load both tag files
    reference_tags = detector.load_tags(options.reference_file)
    current_tags = detector.load_tags(options.current_file)

    # Detect changes
    changes = detector.detect_changes(reference_tags, current_tags)

    # Display changes
    detector.display_changes(changes, options.reference_file,
                            options.current_file, options.verbose)

    # Update reference file if requested
    if options.update_reference:
        detector.update_tags_file(options.reference_file, changes)

    # Exit status:
    #   * Modifications and deletions always fail (reference is stale).
    #   * In strict mode, additions also fail unless --update-reference
    #     absorbed them into the reference file in this run.
    #   * Otherwise, additions are tolerated (legacy behaviour kept for
    #     consumers that do not want CI blocked by new tags).
    additions_remaining = bool(changes.added) and not options.update_reference
    fail = bool(changes.modified or changes.deleted)
    if options.strict and additions_remaining:
        fail = True
    sys.exit(1 if fail else 0)


if __name__ == "__main__":
    main()
